nc-client

gperciva opened the pull request.

cperciva commented on tests/test_spiped.sh:
```diff
@@ -106,7 +106,7 @@ setup_spiped_encryption_server () {
 		-t [127.0.0.1]:${mid_port}	\
 		-p ${s_basename}-spiped-e.pid	\
-		-k /dev/null -o 1
+		-D -k /dev/null -o 1
```
This is weird, and does exactly the opposite of what I think you're trying to do.

gperciva replied:
huh.

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,151 @@
+		goto err2;
+	}
+
+	/* Loop until we die. */
```
This is explicitly not looping until we die.  Rather, it's looping until we're finished with the connection.

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,151 @@
+	exit(0);
+
+err3:
+	network_connect_cancel(connect_cookie);
```
We might arrive here after the connect callback has been invoked.

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,151 @@
+		    == NULL) {
+			warn0("network_write failure");
+			goto err0;
+		};
```
`;`

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,151 @@
+	size_t len = 0;
+	ssize_t read;
+
+	/* Ensure that the buffer is clear. */
```
Since the buffer is guaranteed to be NULL when we first arrive here, why not put the buffer-freeing into callback_wrote?

gperciva commented:
Technically, I didn't need to keep the `write_cookie` -- the only way [1] we'd exit out of the event loop is from `send_input()`, which never has an existing write cookie.  But I figured it was good practice for my knowledge of the libcperciva network API to add it.  :)

[1] other than SIGTERM / SIGKILL, which I'm not going to bother with.

cperciva commented:
On 02/14/17 20:50, Graham Percival wrote:
> Technically, I didn't need to keep the |write_cookie| -- the only way [1] we'd
> exit out of the event loop is from |send_input()|, which never has an existing
> write cookie.

You sure about that?  What if poll(2) fails for some reason?

-- 
Colin Percival
Security Officer Emeritus, FreeBSD | The power to serve
Founder, Tarsnap | www.tarsnap.com | Online backups for the truly paranoid

gperciva commented:
huh, yes, I suppose that the underlying API could have a problem.  Good lesson!

gperciva commented:
if this looks good, I'll rebase the history.

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,157 @@
+{
+	struct senddata * send = (struct senddata *)cookie;
+
+	(void)lenwrit; /* UNUSED */
```
Maybe you want to check `lenwrit`?  If the destination process closes the socket before all the data is sent, do you want nc_client to report an error?

cperciva commented:
Looks fine aside from the one question about `lenwrit`.

gperciva commented:
Updated PR with new error-checking.  And yes, I tried sending my freebsd-i386 `.vdi` to ncat, to ensure that I could kill it during a transmission and get an appropriate error message.  (and have no valgrind errors)

I rebased the previous REBASE commit, but added two more.

cperciva commented on spipe/main.c:
```diff
@@ -176,7 +176,7 @@ main(int argc, char * argv[])
 	}
 
-	/* Loop until we die. */
+	/* Loop until we're done with the connection. */
```
Fine, but it doesn't belong in this PR.

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,175 @@
+};
+
+/* Forward declarations. */
+static int send_input(void *);
```
Not necessary, since the function is defined before it's first used.  I think it's only `callback_wrote` which needs a forward declaration.

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,175 @@
+
+/* Forward declarations. */
+static int send_input(void *);
+static int callback_wrote(void * cookie, ssize_t lenwrit);
```
And you don't need variable names here.

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,175 @@
+			goto err0;
+		}
+	} else {
+		/* Close connection without sending anything. */
```
You might want to check `feof(stdin)` here, to distinguish between EOF and an error?

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,175 @@
+
+	/* Check that we wrote all our data. */
+	if (lenwrit != send->nchars) {
+		warn0("Mismatch in data sent and data requested to send");
```
s/in/between/ ?

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,175 @@
+{
+	struct senddata * send = (struct senddata *)cookie;
+
+	/* We are not longer connecting. */
```
s/not longer/no longer/

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,175 @@
+
+err0:
+	/* Failure! */
+	return (1);
```
We normally return -1 on failure, not 1.  (Although it doesn't really matter.)

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,175 @@
+
+err0:
+	/* Failure! */
+	return (1);
```
We normally return -1 on failure, not 1.  (Although it doesn't really matter.)

cperciva commented on tests/nc-client/main.c:
```diff
@@ -0,0 +1,175 @@
+
+err0:
+	/* Failure! */
+	return (1);
```
We normally return -1 on failure, not 1.  (Although it doesn't really matter.)

gperciva replied:
Whoops, I was thinking about error codes.  Should the `return (1)` in `proto_crypt_dh_validate()` be changed?

cperciva replied:
No, `proto_crypt_dh_validate` isn't returning `1` to mean "an internal error occurred"; it's successfully saying "this is not valid".  Think "true / false / file not found" logic states.

cperciva commented:
Looks good aside from the nit-picking.

gperciva commented:
Updated PR: rebased the previous updates, added more nitpick-fixing, and removed the unaffiliated commit.

cperciva commented:
Looks good.  Rebase and I'll merge once the checks finish.

gperciva commented:
Rebased and tests ok.

gperciva commented:
wait, I just realized it needs another `warn0`.

gperciva commented:
ok, actually ready now.
